client/client.go

```diff
@@ -143,75 +141,102 @@ type PostMultipartOptions struct {
 
 // PostMultipart posts a multipart message in the MIME internet format.
 func (c *Client) PostMultipart(path string, opts PostMultipartOptions, out interface{}) error {
-	body := &bytes.Buffer{}
 
-	writer := multipart.NewWriter(body)
+	r, w := io.Pipe()
+	var pr io.Reader = r
 
-	for name, file := range opts.Files {
-		part, err := writer.CreateFormFile(name, "binary-data")
-		if err != nil {
-			return err
-		}
+	// Get the files size(s) before hand if any
+	if opts.Progress != nil {
 
-		if _, err = io.Copy(part, file); err != nil {
-			return err
+		var size int64
+		for _, file := range opts.Files {
+			if rs, ok := file.(io.ReadSeeker); ok {
+				s, err := io.Copy(ioutil.Discard, rs)
+				if err != nil {
+					return err
+				}
+				size += s
+				rs.Seek(0, 0) // io.SeekStart == 0 in go1.7
+			} else {
+				size = 0 // if even one file isn't seekable, bail
+				break
+			}
 		}
-	}
 
-	for name, value := range opts.Params {
-		writer.WriteField(name, value)
-	}
+		if size > 0 {
 
-	if err := writer.Close(); err != nil {
-		return err
-	}
+			opts.Progress.Start(size)
 
-	br := io.Reader(body)
-
-	if opts.Progress != nil {
-		opts.Progress.Start(int64(body.Len()))
+			defer opts.Progress.Finish()
 
-		defer opts.Progress.Finish()
-
-		br = NewProgressReader(br, opts.Progress.Progress)
+			pr = NewProgressReader(r, opts.Progress.Progress)
+		}
 	}
 
-	req, err := c.request("POST", path, br)
+	writer := multipart.NewWriter(w)
+	streamErr := make(chan error)
+	go func() {
```

Comment:
Since we are using an `io.Pipe` (so that files aren't loaded completely into memory) `CreateFormFile` blocks using the Pipe writer. By having this logic in a goroutine, the request can start reading from the pipe and the goroutine can write to it.